### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by suhan3z**

### CODE REVIEW

File: src/button.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    var _registerEvents = function (button) {
+        var onDragEnd = function (coords) {
+            localStorage.setItem(settings.Constants.BUTTON_POSITION_ITEM_NAME, JSON.stringify(coords));
+        };
+
+        var openMenu = function () {
+            iframeController.setButtonPosition(_getButtonPosition());
+            iframeController.showDetailedMenu();
+        };
+
```
Comment (atropnikov) [inline src/button.js]: Why did you make these functions local?

File: src/button.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+     */
+    var show = function () {
+        if (!_checkRequirements()) {
+            log.info("Environment doesn't satisfy requirements, so don't show Adguard");
+            return;
+        }
+        if (button) {
+            return;
+        }
+        log.info("Requirements checked, all ok");
```
Comment (atropnikov) [inline src/button.js]: Maybe `debug` level is better?

File: src/button.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        button = $(resources.getResource('button.html'));
+        gmApi.GM_addStyle(resources.getResource('button.css'));
+        gmApi.GM_addStyle(resources.getResource('selector.css'));
+        _setPositionSettingsToButton(button);
+        $('body')[0].appendChild(button[0]);
+        _registerEvents(button);
```
Comment (atropnikov) [inline src/button.js]: I don't see `unregisterEvents` function.

File: src/controllers/blockPreviewController.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * @returns {{init: init, setSelectedElement: setSelectedElement}}
+ * @constructor
+ */
+var BlockPreviewController = function ($, selector, gmApi) {
+    var contentDocument = null;
+    var selectedElement = null;
+    var selectedPath = null;
+    var iframeCtrl = Ioc.get('iframeController');
+
+    var init = function (iframe) {
```
Comment (atropnikov) [inline src/controllers/blockPreviewController.js]: Comment

File: src/controllers/blockPreviewController.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            '#end-preview': _showDetailedMenu,
+            '#block-element': _blockElement
+        };
+        Object.keys(menuEvents).forEach(function (item) {
+            $(contentDocument.querySelectorAll(item)).on('click', menuEvents[item]);
+        });
+    };
+
+    var _hideElement = function () {
+        $(selectedElement).addClass('adguard-hide');
```
Comment (atropnikov) [inline src/controllers/blockPreviewController.js]: Is there any error if selectedElement is null?

File: src/button.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
+/**
+ * Adguard assistant button
+ * @param log
```
Comment (ameshkov) [inline src/button.js]: What are those parameters?

File: src/button.js [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        _setPositionSettingsToButton(button);
+        $('body')[0].appendChild(button[0]);
+        _registerEvents(button);
+    };
+
+    /**
+     * Checking browser and other requirements.
+     * @private
+     */
+    var _checkRequirements = function () {
```
Comment (ameshkov) [inline src/button.js]: There's no need in using _lowerCase naming.

Just use lowerCase everywhere

File: src/button.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        for (var i = 0; i < buttonPositionClasses.length; i++) {
+            var currentClass = buttonPositionClasses[i];
+            if (button.hasClass(currentClass)) {
+                button.removeClass(currentClass);
+            }
+        }
+    };
+
+    var _hideRestoreOnFullScreen = function () {
+        if (fullScreenEventsSetted) {
```
Comment (ameshkov) [inline src/button.js]: There's no word `setted`, let it be `Defined` or `Registered`

File: src/controllers/blockPreviewController.js (hunk 5 above)
Comment (ameshkov) [inline src/controllers/blockPreviewController.js]: Agreed, there should be selectedElements validation.

File: src/controllers/mainMenuController.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        gmApi.ADG_temporaryDontBlock(30);
+    };
+
+    var _reportAbuse = function () {
+        gmApi.ADG_sendAbuse();
+        iframeCtrl.removeIframe();
+    };
+
+    var _goToSiteReport = function () {
+        var url = StringUtils.format('https://adguard.com/adguard-report/{0}/report.html', domain);
```
Comment (ameshkov) [inline src/controllers/mainMenuController.js]: Should be defined in the Constants class

File: src/event.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Custom event
+ * @returns {{attach: attach, notify: notify}}
+ * @constructor
+ */
+var Event = function () {
```
Comment (ameshkov) [inline src/event.js]: There is an `Event` class already, you'd better change the class name

File: src/ioc.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+"use strict";
+/**
+ * Simple ioc implementation
+ * @type {{add, get}}
+ */
+var Ioc = (function () {
+    var dependencies = {};
+
+    var add = function (qualifier, obj) {
```
Comment (ameshkov) [inline src/ioc.js]: It does not "adds" an object, you should call it either "register" or "set"

File: src/less/blocking-preview.php [hunk 12]
```diff
@@ -0,0 +1,34 @@
+<? include "header.php"; ?>
```
Comment (ameshkov) [inline src/less/blocking-preview.php]: What does this file do here?

File: src/less/blocking-rule.php [hunk 13]
```diff
@@ -0,0 +1,117 @@
+<? include "header.php"; ?>
```
Comment (ameshkov) [inline src/less/blocking-rule.php]: The same question

File: src/less/blocking.php [hunk 14]
```diff
@@ -0,0 +1,27 @@
+<? include "header.php"; ?>
```
Comment (ameshkov) [inline src/less/blocking.php]: Again

File: src/less/footer.php [hunk 15]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (ameshkov) [inline src/less/footer.php]: Again

File: src/localization.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
+var SupportedLocales = {'en': en, 'ru': ru};
```
Comment (atropnikov) [inline src/localization.js]: Make it local

File: src/iframe.js [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+        body.appendChild(view);
+    };
+
+    var setCloseEventIfNotHitIframe = function (isNeedToBeSet) {
+        if (!needSetCloseEvent) {
+            return;
+        }
+        if (isNeedToBeSet && !document.onclick) {
+            window.setTimeout(function () {
+                document.onclick = function () {
```
Comment (atropnikov) [inline src/iframe.js]: Strange code. Add `click` event listener with $.

File: src/main.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
+Ioc.register('log', new Log());
```
Comment (atropnikov) [inline src/main.js]: Wrap this code into anonymous function.

File: src/settings.js [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        Constants: Constants,
+        MenuItemsNames: MenuItemsNames,
+        getSettings: getSettings,
+        loadSettings: loadSettings,
+        getWotData: getWotData,
+        setWotData: setWotData,
+        saveSettings: saveSettings,
+        getUserPositionForButton: getUserPositionForButton,
+        removeUserPosition: removeUserPosition,
+        selectedElement: selectedElement,
```
Comment (atropnikov) [inline src/settings.js]: Unused

File: src/settings.js [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+    var getAdguardSettings = function () {
+        return adguardSettings;
+    };
+
+    var removeUserPosition = function () {
+        localStorage.removeItem(Constants.BUTTON_POSITION_ITEM_NAME);
+    };
+
+    var setUserPosition = function (coords) {
+        localStorage.setItem(settings.Constants.BUTTON_POSITION_ITEM_NAME, JSON.stringify(coords));
```
Comment (atropnikov) [inline src/settings.js]: `settings` isn't defined.

File: src/settings.js [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+        MenuItemsNames: MenuItemsNames,
+        getSettings: getSettings,
+        loadSettings: loadSettings,
+        getWotData: getWotData,
+        setWotData: setWotData,
+        saveSettings: saveSettings,
+        getUserPositionForButton: getUserPositionForButton,
+        removeUserPosition: removeUserPosition,
+        setUserPosition: setUserPosition,
+        selectedElement: selectedElement,
```
Comment (atropnikov) [inline src/settings.js]: Remove `selectedElement`

File: src/settings.js [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+    };
+
+    var getUserPositionForButton = function () {
+        try {
+            var userPosition = localStorage.getItem(Constants.BUTTON_POSITION_ITEM_NAME);
+            log.info("Check user position for domain");
+            if (userPosition) {
+                log.info("User position is set for this domain");
+                var position = JSON.parse(userPosition);
+                if (!position.top || !position.left) {
```
Comment (atropnikov) [inline src/settings.js]: `!0 === true`

File: src/settings.js (hunk 21 above)
Comment (atropnikov) [inline src/settings.js]: removeUserPosition => removeUserPositionForButton
setUserPosition => setUserPositionForButton

File: src/localization.js [hunk 23]
```diff
@@ -0,0 +1,97 @@
+/**
+ * Object that manages localizations
+ * @returns {{getMessage: Function, translateElement: Function}}
+ * @constructor
+ */
+var Localization = function () {
+    var currentLocale = null;
+    var SupportedLocales = {'en': en, 'ru': ru};
+
+    if (typeof AdguardSettings != 'undefined') {
```
Comment (atropnikov) [inline src/localization.js]: Use `settings.getAdguardSettings`

File: src/localization.js [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+            currentLocale = locale;
+        }
+    }
+    if (!currentLocale) {
+        currentLocale = 'en';
+    }
+
+
+    var getMessage = function (messageId) {
+        return SupportedLocales[currentLocale][messageId].message;
```
Comment (atropnikov) [inline src/localization.js]: What if `messageId` isn't in `SupportedLocales[currentLocale]`

File: src/ioc.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+
+    var register = function (qualifier, obj) {
+        dependencies[qualifier] = obj;
+    };
+
+    var get = function (func) {
+        if (typeof func == 'string') {
+            var resolved = dependencies[func];
+            if (!resolved) {
+                throw StringUtils.format("Can't resolve {0}", func);
```
Comment (atropnikov) [inline src/ioc.js]: Simply `"Can't resolve " + func`

File: src/iframe.js [hunk 26]
```diff
@@ -0,0 +1,277 @@
...
+        $(iframe).on('load', function () {
+            if (iframeAlreadyLoaded) {
+                //IE calls load each time when we use document.close
+                return;
+            }
+            iframeAlreadyLoaded = true;
+            appendDefaultStyle();
+            onIframeLoadCallback();
+        });
+        $('body')[0].appendChild(iframe[0]);
```
Comment (atropnikov) [inline src/iframe.js]: `body` may not exist

File: src/iframe.js [hunk 27]
```diff
@@ -0,0 +1,277 @@
...
+        }
+
+    };
+
+    var appendDefaultStyle = function () {
+        try {
+            log.info('Iframe loaded writing styles');
+            var doc = iframe[0].contentDocument;
+            doc.open();
+            doc.write(StringUtils.format("<html><head>{0}</head></html>",
```
Comment (atropnikov) [inline src/iframe.js]: Be consistent. Either String.format or concatenation.

File: src/button.js [hunk 28]
```diff
@@ -0,0 +1,168 @@
...
+        }
+        if (button) {
+            return;
+        }
+        log.debug("Requirements checked, all ok");
+        button = $(resources.getResource('button.html'));
+        gmApi.GM_addStyle(resources.getResource('button.css'));
+        gmApi.GM_addStyle(resources.getResource('selector.css'));
+        setPositionSettingsToButton(button);
+        $('body')[0].appendChild(button[0]);
```
Comment (atropnikov) [inline src/button.js]: `body` may not exist

File: src/button.js [hunk 29]
```diff
@@ -0,0 +1,168 @@
...
+                button.removeClass(currentClass);
+            }
+        }
+    };
+
+    var hideRestoreOnFullScreen = function () {
+        if (isFullScreenEventsRegistered) {
+            return;
+        }
+        $(document).on('webkitfullscreenchange mozfullscreenchange fullscreenchange', function () {
```
Comment (atropnikov) [inline src/button.js]: I don't see removing this event listener.

File: src/button.js [hunk 30]
```diff
@@ -0,0 +1,168 @@
...
+            if (uiUtils.tryFullScreenPrefix(document, "FullScreen") || uiUtils.tryFullScreenPrefix(document, "IsFullScreen")) {
+                hideButton();
+            } else {
+                showButton();
+            }
+        });
+        isFullScreenEventsRegistered = true;
+    };
+
+    var hideButton = function () {
```
Comment (atropnikov) [inline src/button.js]: It had better add additional check that button isn't null.

File: src/utils/common-utils.js [hunk 31]
```diff
@@ -0,0 +1,201 @@
+/**
+ * String utils
+ */
+var StringUtils = {
+    /**
+     * Replaces the format items in a specified String with the text equivalents of the values of corresponding object instances.
+     * @param format
+     * @param args
+     */
+    format: function (format, args) {
```
Comment (atropnikov) [inline src/utils/common-utils.js]: `args` is unused.

File: src/utils/common-utils.js [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+            var child;
+            var i;
+            for (i = 0; i < children.length; i++) {
+                if (children[i].nodeType == 1) {
+                    child = children[i];
+                    count++;
+                }
+            }
+        }
+        return count == 1 ? child : null;
```
Comment (atropnikov) [inline src/utils/common-utils.js]: `count` is used out of scope.

File: src/utils/ui-utils.js [hunk 33]
```diff
@@ -0,0 +1,229 @@
...
+    var makeElementDraggable = function (element, onDragEnd, onClick, onMouseDown) {
+        var getCoords = function (elem) {
+            var box = elem.getBoundingClientRect();
+            return {
+                top: box.top,
+                left: box.left
+            };
+        };
+
+        element.onmousedown = function (e) {
```
Comment (atropnikov) [inline src/utils/ui-utils.js]: `$(element).on('mousedown', ...`

File: src/utils/ui-utils.js [hunk 34]
```diff
@@ -0,0 +1,229 @@
...
+                element.style.left = e.pageX - shiftX + 'px';
+                element.style.top = e.pageY - shiftY + 'px';
+            };
+
+            moveAt(e);
+            if (onMouseDown) {
+                onMouseDown();
+            }
+
+            document.onmousemove = function (e) {
```
Comment (atropnikov) [inline src/utils/ui-utils.js]: `$(document).on('mousemove', ...` + event listener cleanup

Comment (atropnikov) [inline src/utils/ui-utils.js]: Also, it would be great use `debounce` here.

File: src/utils/ui-utils.js [hunk 35]
```diff
@@ -0,0 +1,229 @@
...
+            $iframeDocument.off('mousemove', onMouseMove);
+            $iframeDocument.off('selectstart', cancelIFrameSelection);
+        };
+
+        dragHandle.on('mousedown', onMouseDown);
+        $iframeDocument.on('mouseup', onMouseUp);
+    };
+
+    var browser_prefixes = ["webkit", "moz", "ms", "o", ""];
+    var tryFullScreenPrefix = function (obj, method) {
```
Comment (atropnikov) [inline src/utils/ui-utils.js]: Comment

File: src/utils/ui-utils.js [hunk 36]
```diff
@@ -0,0 +1,229 @@
...
+        var onMouseUp = function () {
+            $iframeDocument.off('mousemove', onMouseMove);
+            $iframeDocument.off('selectstart', cancelIFrameSelection);
+        };
+
+        dragHandle.on('mousedown', onMouseDown);
+        $iframeDocument.on('mouseup', onMouseUp);
+    };
+
+    var browser_prefixes = ["webkit", "moz", "ms", "o", ""];
```
Comment (atropnikov) [inline src/utils/ui-utils.js]: Bad variable naming

File: src/utils/ui-utils.js [hunk 37]
```diff
@@ -0,0 +1,229 @@
...
+    var validateBrowser = function () {
+        return !document.documentMode || (document.documentMode > settings.Constants.MINIMUM_IE_SUPPORTED_VERSION);
+    };
+
+    /**
+     * Checks if page is valid for Adguard assistant to work here.
+     */
+    var validatePage = function () {
+        // Assistant do not work in iframes
+        if (window != top) return false;
```
Comment (atropnikov) [inline src/utils/ui-utils.js]: Try to use strict comparison where it's possible. Also, I see JSHint warnings here.

File: src/controllers/blockPreviewController.js [hunk 38]
```diff
@@ -0,0 +1,71 @@
+/**
+ * Block preview controller
+ * @param $
+ * @param selector
+ * @param gmApi
+ * @returns {{init: init, setSelectedElement: setSelectedElement}}
+ * @constructor
+ */
+var BlockPreviewController = function ($, selector, gmApi) {
```
Comment (atropnikov) [inline src/controllers/blockPreviewController.js]: I see JSHint warnings in this file.

**Pull request merged**